File: engine/src/input_mapper.cpp
#include "input_mapper.h"
#include <memory>
#include "script_wrappers.h"

namespace fe_engine {
	input_mapper::input_mapper(reference<controller> controller) {
		m_controller = controller;
		script_wrappers::set_imapper(reference<input_mapper>(this));
	}
	void input_mapper::update()
	{
		m_current = commands({ 0 });
		if (m_controller) {
			m_controller->update();
			auto buttons = m_controller->get_state();

			if (buttons.up.down) {
				m_current.up = true;
			}
			if (buttons.down.down) {
				m_current.down = true;
			}
			if (buttons.left.down) {
				m_current.left = true;
			}
			if (buttons.right.down) {
				m_current.right = true;
			}
			if (buttons.a.down) {
				m_current.ok = true;
			}
			if (buttons.b.down) {
				m_current.back = true;
			}
			if (buttons.start.down) {
				m_current.exit = true;
			}
		}
	}
}

File: scriptcore/src/InputMapper.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using System.Runtime.CompilerServices;
using FEEngine.Util;

namespace FEEngine
{
    public class InputMapper : RegisteredObject<InputMapper>
    {
        public struct Commands
        {
            public bool Up;
            public bool Down;
            public bool Left;
            public bool Right;
            public bool OK;
            public bool Back;
            public bool Exit;
        }
        public Commands GetState()
        {
            return GetState_Native(this.Index);
        }
        public ulong Index { get; private set; }
        public InputMapper()
        {
            this.Index = 0;
        }
        public void SetRegisterIndex(ulong index)
        {
            this.Index = index;
        }
        public static InputMapper MakeFromRegisterIndex(ulong index)
        {
            InputMapper im = new InputMapper();
            im.Index = index;
            return im;
        }
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern Commands GetState_Native(ulong index);
    }
}
